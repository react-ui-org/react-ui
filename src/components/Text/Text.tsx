import React from 'react';
import { withGlobalProps } from '../../providers/globalProps';
import { classNames } from '../../helpers/classNames';
import { transferProps } from '../../helpers/transferProps';
import { isChildrenEmpty } from '../_helpers/isChildrenEmpty';
import { getRootClampClassName } from './_helpers/getRootClampClassName';
import { getRootHyphensClassName } from './_helpers/getRootHyphensClassName';
import { getRootWordWrappingClassName } from './_helpers/getRootWordWrappingClassName';
import styles from './Text.module.scss';
import { TextProps } from './Text.types';

export const Text: React.FunctionComponent<TextProps> = ({
  blockLevel = false,
  children,
  hyphens = 'none',
  lines,
  wordWrapping = 'normal',
  ...restProps
}) => {
  if (isChildrenEmpty(children)) {
    return null;
  }

  const HtmlElement = blockLevel ? 'div' : 'span';

  return (
    <HtmlElement
      {...transferProps(restProps)}
      className={(hyphens !== 'none' || (lines && lines > 0) || wordWrapping !== 'normal')
        ? classNames(
          getRootClampClassName(styles, lines),
          getRootHyphensClassName(hyphens, styles),
          getRootWordWrappingClassName(wordWrapping, styles),
        )
        : undefined}
      style={(lines && lines > 1) ? { '--rui-custom-lines': lines } as React.CSSProperties : undefined}
    >
      {children}
    </HtmlElement>
  );
};

export const TextWithGlobalProps = withGlobalProps(Text, 'Text');

export default TextWithGlobalProps;
