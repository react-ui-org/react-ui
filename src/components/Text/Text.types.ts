import { ReactNode } from 'react';
import {
  Hyphens,
  WordWrapping,
} from '../../types';

export type TextProps = React.ComponentProps<'div' | 'span'> & {
  /**
   * If true, the root HTML element renders as `<div>` instead of `<span>`.
   */
  blockLevel?: boolean;
  /**
   * Text content to be sanitized. Can contain HTML.
   */
  children?: ReactNode;
  /**
   * Turn on hyphenation. Head to [Hyphens](#hyphens) to learn more.
   */
  hyphens?: Hyphens;
  /**
   * Optional number of lines. If exceeded, the content is truncated and appended by an ellipsis (`â€¦`).
   */
  lines?: number;
  /**
   * How to deal with long words. Head to [Word Wrapping](#word-wrapping) for detailed explanation.
   */
  wordWrapping?: WordWrapping;
};
