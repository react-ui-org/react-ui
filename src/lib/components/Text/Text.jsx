import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootHyphensClassName } from './_helpers/getRootHyphensClassName';
import { getRootWordWrappingClassName } from './_helpers/getRootWordWrappingClassName';
import styles from './Text.scss';

export const Text = ({
  blockLevel,
  children,
  hyphens,
  id,
  lines,
  wordWrapping,
}) => {
  const HtmlElement = blockLevel ? 'div' : 'span';

  return (
    <HtmlElement
      className={(hyphens !== 'none' || lines > 0 || wordWrapping !== 'normal')
        ? classNames(
          (lines > 0) ? styles.rootClampLines : '',
          getRootHyphensClassName(hyphens, styles),
          getRootWordWrappingClassName(wordWrapping, styles),
        )
        : undefined}
      id={id}
      style={(lines > 0) ? { '--rui-custom-lines': lines } : undefined}
    >
      {children}
    </HtmlElement>
  );
};

Text.defaultProps = {
  blockLevel: false,
  children: null,
  hyphens: 'none',
  id: undefined,
  lines: undefined,
  wordWrapping: 'normal',
};

Text.propTypes = {
  /**
   * If true, the root HTML element renders as `<div>` instead of `<span>`.
   */
  blockLevel: PropTypes.bool,
  /**
   * Text content to be sanitized. Can contain HTML.
   */
  children: PropTypes.node,
  /**
   * Turn on hyphenation. Head to [Hyphens](#hyphens) to learn more.
   */
  hyphens: PropTypes.oneOf(['none', 'auto', 'manual']),
  /**
   * Optional ID of the root HTML element.
   */
  id: PropTypes.string,
  /**
   * Optional number of lines. If exceeded, the content is truncated and appended by an ellipsis (`â€¦`).
   */
  lines: PropTypes.number,
  /**
   * How to deal with long words. Head to [Word Wrapping](#word-wrapping) for detailed explanation.
   */
  wordWrapping: PropTypes.oneOf(['normal', 'long-words', 'anywhere']),
};

export const TextWithContext = withProviderContext(Text, 'Text');

export default TextWithContext;
