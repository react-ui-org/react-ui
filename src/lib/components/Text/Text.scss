// 1. `word-break: break-word` is deprecated in favour of `overflow-wrap: anywhere`, but it's still
//     required for Safari.
//     https://caniuse.com/mdn-css_properties_overflow-wrap_anywhere

// stylelint-disable property-no-vendor-prefix, value-no-vendor-prefix
.rootClampLines {
    display: -webkit-box;
    -webkit-line-clamp: var(--rui-custom-lines);
    -webkit-box-orient: vertical;
    overflow: hidden;
    text-overflow: ellipsis;
}
// stylelint-enable property-no-vendor-prefix, value-no-vendor-prefix

.rootHyphensAuto {
    hyphens: auto;
}

.rootHyphensManual {
    hyphens: manual;
}

.rootWordWrappingAnywhere {
    word-break: break-all;
}

.rootWordWrappingLongWords {
    word-break: break-word; // 1.
    overflow-wrap: anywhere;
}
