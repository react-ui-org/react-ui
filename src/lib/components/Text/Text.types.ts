export interface TextProps {
  /**
   * Rest of the props.
   */
  [key: string]: unknown;
  /**
   * If true, the root HTML element renders as `<div>` instead of `<span>`.
   */
  blockLevel?: boolean,
  /**
   * Text content to be sanitized. Can contain HTML.
   */
  children?: React.ReactNode,
  /**
   * Turn on hyphenation. Head to [Hyphens](#hyphens) to learn more.
   */
  hyphens?: Hyphens,
  /**
   * Optional number of lines. If exceeded, the content is truncated and appended by an ellipsis (`â€¦`).
   */
  lines?: number,
  /**
   * How to deal with long words. Head to [Word Wrapping](#word-wrapping) for detailed explanation.
   */
  wordWrapping?: WordWrapping
}
