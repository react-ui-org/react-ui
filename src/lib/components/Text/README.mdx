---
name: Text
menu: 'Miscellaneous'
route: /components/text
---

# Text

import {
  Playground,
  Props,
} from 'docz'
import { Placeholder } from '../../../docs/_components/Placeholder/Placeholder'
import { Toolbar } from '../Toolbar/Toolbar'
import { ToolbarGroup } from '../Toolbar/ToolbarGroup'
import { ToolbarItem } from '../Toolbar/ToolbarItem'
import { Button } from '../Button/Button'
import { ButtonGroup } from '../ButtonGroup/ButtonGroup'
import { TextField } from '../TextField/TextField'
import { Text } from './Text'

Text is a tiny component to control wrapping of text content.

## Basic Usage

To implement the Text component, you need to import it first:

```js
import { Text } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <Placeholder>
    <Text lines={3}>
      Hello! This text will be clamped to three lines.
      Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo
      ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis
      dis parturient montes, nascetur ridiculus mus. Donec quam felis,
      ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa
      quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget,
      arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo.
    </Text>
  </Placeholder>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use Text anytime you need to **control potentially long strings** and
  prevent them from overflowing or breaking their container.

- **By default, Text doesn't alter rendering of its content.** Use available
  options to achieve the result you need.

- **Text renders as `<span>` by default,** so it can only contain
  inline-level HTML elements (like `<strong>` or `<a>`, but not `<div>`, `<p>`,
  or `<h2>`). If necessary, use the `blockLevel` option to render as `<div>` to
  keep your HTML valid.

- **Use Text for short pieces of text content.** Including a couple of HTML tags
  shouldn't cause any harm, but keep in mind Text is not intended to wrap
  complex HTML structures or even React components.

## Number of Lines

Specify how many `lines` of text should be visible if content doesn't fit its
container. If the number of lines is exceeded, the content is truncated and
appended by an ellipsis (`â€¦`).

<Playground>
  {() => {
    const [lines, setLines] = React.useState(1);
    return (
      <div>
        <Toolbar align="baseline">
          <ToolbarItem>
            <TextField
              label="Number of lines"
              min={1}
              max={100}
              onChange={(e) => setLines(e.target.value)}
              type="number"
              value={lines}
            />
          </ToolbarItem>
        </Toolbar>
        <Placeholder>
          <Text lines={lines}>
            Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean
            commodo ligula eget dolor. Aenean massa. Cum sociis natoque
            penatibus et magnis dis parturient montes, nascetur ridiculus mus.
            Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem.
            Nulla consequat massa quis enim. Donec pede justo, fringilla vel,
            aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut,
            imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede
            mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum
            semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula,
            porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem
            ante, dapibus in, viverra quis, feugiat a, tellus.
          </Text>
        </Placeholder>
      </div>
    );
  }}
</Playground>

## Word Wrapping

There are three possible ways of controlling wrapping of long words if they do
not fit on a line. Set `wordWrapping` to one of the following values:

- `normal`: Do not force any wrapping (default behavior).

- `long-words`: To prevent overflow, an otherwise unbreakable string of
  characters â€” like a long word or URL â€” may be broken at any point if there are
  no otherwise-acceptable break points in the line.

- `anywhere`: Create a break at the exact place where text would otherwise
  overflow its container (even if putting an entire word on its own line would
  negate the need for a break).

ðŸ“– [Read more about wrapping and breaking text at MDN.](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Text/Wrapping_Text)

### Hyphens

The `hyphens` option specifies how words should be hyphenated when text wraps
across multiple lines. It can prevent hyphenation entirely, hyphenate at
manually-specified points within the text, or let the browser automatically
insert hyphens where appropriate.

ðŸ‘‰ The `auto` setting's behavior depends on the language being properly tagged
to select the appropriate hyphenation rules. **You must specify a language**
using the `lang` HTML attribute (e.g.
[on `<html>` tag](/getting-started/usage#full-example)) to guarantee that
automatic hyphenation is applied in that language.

ðŸ‘‰ [Manually suggested line break opportunities](https://developer.mozilla.org/en-US/docs/Web/CSS/hyphens#suggesting_line_break_opportunities)
will override automatic break point selection in `auto` mode when present.

ðŸ“– [Read more about `hyphens` CSS property at MDN.](https://developer.mozilla.org/en-US/docs/Web/CSS/hyphens)

<Playground>
  {() => {
    const [hyphens, setHyphens] = React.useState('none');
    const [wordWrapping, setWordWrapping] = React.useState('normal');
    return (
      <div>
        <Toolbar>
          <ToolbarGroup align="baseline">
            <ToolbarItem>
              <span id="word-wrapping-options-label">Word wrapping:</span>
            </ToolbarItem>
            <ToolbarItem>
              <ButtonGroup aria-labelledby="#word-wrapping-options-label">
                <Button
                  color={wordWrapping === 'normal' ? 'dark' : 'primary'}
                  label="normal"
                  onClick={() => setWordWrapping('normal')}
                />
                <Button
                  color={wordWrapping === 'long-words' ? 'dark' : 'primary'}
                  label="long-words"
                  onClick={() => setWordWrapping('long-words')}
                />
                <Button
                  color={wordWrapping === 'anywhere' ? 'dark' : 'primary'}
                  label="anywhere"
                  onClick={() => setWordWrapping('anywhere')}
                />
              </ButtonGroup>
            </ToolbarItem>
          </ToolbarGroup>
          <ToolbarGroup align="baseline">
            <ToolbarItem>
              <span id="hyphens-options-label">Hyphens:</span>
            </ToolbarItem>
            <ToolbarItem>
              <ButtonGroup aria-labelledby="#hyphens-options-label">
                <Button
                  color={hyphens === 'none' ? 'dark' : 'primary'}
                  label="none"
                  onClick={() => setHyphens('none')}
                />
                <Button
                  color={hyphens === 'auto' ? 'dark' : 'primary'}
                  label="auto"
                  onClick={() => setHyphens('auto')}
                />
                <Button
                  color={hyphens === 'manual' ? 'dark' : 'primary'}
                  label="manual"
                  onClick={() => setHyphens('manual')}
                />
              </ButtonGroup>
            </ToolbarItem>
          </ToolbarGroup>
        </Toolbar>
        <Placeholder width="11em" bordered>
          <Text hyphens={hyphens} wordWrapping={wordWrapping}>
            {hyphens === 'manual'
              ? (<>LongWord&shy;ThatHasManual&shy;Breaking&shy;Possibilities</>)
              : (<>LongWordThatHasNoBreakingPossibilities</>)}
            {' '}
            and a couple of ordinary words that are nice and well behaved.
          </Text>
        </Placeholder>
      </div>
    );
  }}
</Playground>

## API

<Props table of={Text} />
