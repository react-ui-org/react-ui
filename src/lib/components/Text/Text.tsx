import React, { CSSProperties } from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import { classNames } from '../../utils/classNames';
import { isChildrenEmpty } from '../_helpers/isChildrenEmpty';
import { getRootClampClassName } from './_helpers/getRootClampClassName';
import { getRootHyphensClassName } from './_helpers/getRootHyphensClassName';
import { getRootWordWrappingClassName } from './_helpers/getRootWordWrappingClassName';
import styles from './Text.scss';
import { TextProps } from './Text.types';

export const Text: React.FunctionComponent<TextProps> = ({
  blockLevel = false,
  children = null,
  hyphens = 'none',
  lines,
  wordWrapping = 'normal',
  ...restProps
}) => {
  if (isChildrenEmpty(children)) {
    return null;
  }

  const HtmlElement = blockLevel ? 'div' : 'span';

  return (
    <HtmlElement
      {...transferProps(restProps)}
      className={(hyphens !== 'none' || (lines && lines > 0) || wordWrapping !== 'normal')
        ? classNames(
          // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
          getRootClampClassName(lines!, styles),
          getRootHyphensClassName(hyphens, styles),
          getRootWordWrappingClassName(wordWrapping, styles),
        )
        : undefined}
      style={(lines && lines > 1) ? { '--rui-custom-lines': lines } as CSSProperties : undefined}
    >
      {children}
    </HtmlElement>
  );
};

export const TextWithGlobalProps = withGlobalProps(Text, 'Text');

export default TextWithGlobalProps;
