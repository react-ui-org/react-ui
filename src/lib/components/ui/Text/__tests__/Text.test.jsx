import React from 'react';
import { shallow } from 'enzyme';
import { Text } from '../Text';

describe('rendering', () => {
  it('renders correctly with no props', () => {
    const tree = shallow(
      <Text>
        test
      </Text>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly as div', () => {
    const tree = shallow(
      <Text blockLevel>
        test
      </Text>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with word wrapping', () => {
    const tree = shallow(
      <Text wordWrapping="long-words">
        test
      </Text>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with line clamp', () => {
    const tree = shallow(
      <Text lines={2}>
        test
      </Text>,
    );

    expect(tree).toMatchSnapshot();
  });
});
