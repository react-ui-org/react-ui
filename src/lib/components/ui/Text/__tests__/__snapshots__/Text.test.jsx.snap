// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly as div 1`] = `
<div>
  test
</div>
`;

exports[`rendering renders correctly with line clamp 1`] = `
<span
  className="rootClampLines  "
  style={
    Object {
      "--rui-custom-lines": 2,
    }
  }
>
  test
</span>
`;

exports[`rendering renders correctly with no props 1`] = `
<span>
  test
</span>
`;

exports[`rendering renders correctly with word wrapping 1`] = `
<span
  className="  rootWordWrappingLongWords"
>
  test
</span>
`;
